Dependencies/xr/Source/OpenXR/XR.cpp

```diff
@@ -276,57 +308,113 @@ namespace xr
+                for (const auto& viewConfigType : HmdImpl.SupportedSecondaryViewConfigurationTypes)
+                {
+                    const auto& state = RenderResources.ViewStates.at(viewConfigType);
+                    if (state.Views.size() <= secondaryViewIndex)
+                    {
+                        secondaryViewIndex -= state.Views.size();
+                        continue;
+                    }
 
-    private:
-        static constexpr XrPosef IDENTITY_TRANSFORM{ XrQuaternionf{ 0.f, 0.f, 0.f, 1.f }, XrVector3f{ 0.f, 0.f, 0.f } };
+                    const auto& viewConfig = state.ViewConfigViews.at(secondaryViewIndex);
+                    const Size viewSize = { viewConfig.recommendedImageRectWidth, viewConfig.recommendedImageRectHeight };
+                    if (RenderResources.ColorSwapchains.count(viewConfigType) > 0 &&
+                        RenderResources.ColorSwapchains.at(viewConfigType).size() > secondaryViewIndex)
+                    {
+                        const auto& swapchain = RenderResources.ColorSwapchains.at(viewConfigType).at(secondaryViewIndex);
+                        assert(swapchain.Width == viewSize.Width && swapchain.Height == viewSize.Height);
+                    }
 
-        void InitializeRenderResources(XrInstance instance, XrSystemId systemId)
+                    return viewSize;
+                }
+            }
+
+            throw std::exception{/*unknown view index*/ };
+        }
+
+        void PopulateSwapchains(ViewConfigurationState& viewState)
         {
-            // Read graphics properties for preferred swapchain length and logging.
-            XrSystemProperties systemProperties{ XR_TYPE_SYSTEM_PROPERTIES };
-            XrCheck(xrGetSystemProperties(instance, systemId, &systemProperties));
+            assert(viewState.Views.size() > 0);
 
             // Select color and depth swapchain pixel formats
             SwapchainFormat colorSwapchainFormat;
             SwapchainFormat depthSwapchainFormat;
             SelectSwapchainPixelFormats(colorSwapchainFormat, depthSwapchainFormat);
 
-            // Query and cache view configuration views. Two-call idiom.
-            uint32_t viewCount;
-            XrCheck(xrEnumerateViewConfigurationViews(instance, systemId, HmdImpl.VIEW_CONFIGURATION_TYPE, 0, &viewCount, nullptr));
-            assert(viewCount == HmdImpl.STEREO_VIEW_COUNT);
-            RenderResources.ConfigViews.resize(viewCount, { XR_TYPE_VIEW_CONFIGURATION_VIEW });
-            XrCheck(xrEnumerateViewConfigurationViews(instance, systemId, HmdImpl.VIEW_CONFIGURATION_TYPE, viewCount, &viewCount, RenderResources.ConfigViews.data()));
-
-            // Create all the swapchains.
+            uint32_t viewCount = static_cast<uint32_t>(viewState.Views.size());
+            RenderResources.ColorSwapchains[viewState.Type] = {};
+            RenderResources.DepthSwapchains[viewState.Type] = {};
```

Comment:
nit: Will this construct, then copy? If so, is it better to move the vector in, or construct it in-place somehow?

Reply:
this should now be fixed.